Expose fieldName when able in deserialization errors

dcheckoway opened the pull request:
The use case for this PR is a REST API whose error responses must identify which field is missing or malformed.

The change is simple and backward compatible -- change `DeserializationException` to a case class with an optional `fieldName`, and add an optional `fieldName` arg to `deserializationError()`.  This allows apps to do things like this:

```scala
  implicit def rejectionHandler = RejectionHandler {
    case MalformedRequestContentRejection(message, cause) :: _ =>
      cause match {
        case DeserializationException(_, _, Some(fieldName)) =>
          ...the error response can be specific about fieldName
```

fommil commented:
btw, we have clean errors like this automatically in ENSIME Jerk https://twitter.com/fommil/status/589828529904291840

sirthias commented:
This looks good! Thanks, Dan!
Could you accept the [Typesafe CLA](http://www.typesafe.com/contribute/cla) so we can get this merged?

sirthias commented on src/main/scala/spray/json/package.scala:
```diff
@@ -32,7 +32,7 @@ package object json {
 package json {
 
-  class DeserializationException(msg: String, cause: Throwable = null) extends RuntimeException(msg, cause)
+  case class DeserializationException(msg: String, cause: Throwable = null, fieldName: Option[String] = None) extends RuntimeException(msg, cause)
```
How about we turn `fieldName` into `fieldNames: List[String]` so we can gather the *path* to the error field rather than just the inner most member name?

fommil replied:
interesting that you want to have the `fieldNames` in the exception. I find that having the type that is being serialised is more general. Perhaps have both? I resort to including this info in a string, obviously better to have exception level support https://github.com/fommil/ensime-server/blob/jerk/spray-json-shapeless/src/main/scala/org/ensime/json/package.scala#L6

sirthias replied:
I agree that getting the type info would be another great thing to have.
However, while gathering up the field names is an easy, noninvasive thing we can do here, including the type info is a bit more difficult to do properly.

sirthias commented on src/main/scala/spray/json/ProductFormats.scala:
```diff
@@ -56,9 +56,11 @@ trait ProductFormats extends ProductFormatsInstances {
-          deserializationError("Object is missing required member '" + fieldName + "'", e)
+          deserializationError("Object is missing required member '" + fieldName + "'", e, fieldName)
+        case DeserializationException(msg, cause, None) =>
+          deserializationError(msg, cause, fieldName)
```
This would then become:
case DeserializationException(msg, cause, fieldNames) =>
  deserializationError(msg, cause, fieldName :: fieldNames)
```scala

```

dcheckoway commented:
@sirthias I appreciate the review & feedback.  I've changed it to `fieldNames: List` and added a bit of test coverage for the "gathering" aspect.  I'll email you the signed Typesafe CLA (or let me know if I should send it elsewhere).  Thanks!

sirthias commented:
@jrudolph Do you have any comments here?
